node/core/dispute-coordinator/src/participation/queues/mod.rs

```diff
@@ -231,15 +226,31 @@ impl Queues {
 		&mut self,
 		comparator: CandidateComparator,
 		priority: ParticipationPriority,
-		req: ParticipationRequest,
+		mut req: ParticipationRequest,
 	) -> std::result::Result<(), QueueError> {
 		if priority.is_priority() {
 			if self.priority.len() >= PRIORITY_QUEUE_SIZE {
 				return Err(QueueError::PriorityFull)
 			}
-			// Remove any best effort entry:
-			self.best_effort.remove(&comparator);
-			self.priority.insert(comparator, req);
+			// Remove any best effort entry, using it to replace our new
+			// request. This preserves the original request timer from
+			// the older request.
+			if let Some(older_request) = self.best_effort.remove(&comparator) {
+				if let Some(timer) = req.request_timer {
+					timer.stop_and_discard();
+				}
+				req = older_request;
+			}
+			// Keeping old request if any. This prevents request timers
+			// from resetting on each new request at the same priority
+			// level for the same candidate.
+			if self.priority.contains_key(&comparator) {
+				if let Some(timer) = req.request_timer {
+					timer.stop_and_discard();
```

Comment:
Doesn't this stop an old request from the best effort queue? Not sure I'm totally following the logic though -- I think a docstring for the function would help me. 🙂

Reply:
I think you're reading the code right. If there were identical old requests in both the best effort and priority queues then the old best effort request would be thrown out.

But in practice a request should never be in both the best effort queue and the priority queue at the same time. We remove best effort queue entries to ensure this.

If by some fluke a request ends up in both queues it wouldn't be the end of the world. We still get decent data, since we still have the property that no timers send metric data before we've participated in the dispute.